app-backend/datamodel/src/main/scala/com/azavea/rf/datamodel/ColorCorrect.scala

```diff
@@ -111,6 +117,216 @@ object SaturationAdjust {
   }
 }
 
+case class Memo[I <% K, K, O](f: I => O) extends (I => O) {
+  import collection.mutable.{Map => Dict}
+  type Input = I
+  type Key = K
+  type Output = O
+
+  val cache = Dict.empty[K, O]
+  override def apply(x: I) = cache.getOrElseUpdate(x, f(x))
+}
+
+object WhiteBalance {
+  
+  @inline def clamp8Bit(z: Int): Int = {
+    if (z < 0) 0
+    else if (z > 255) 255
+    else z
+  }
+
+  def stitch(pieces: List[(MultibandTile, (Int, Int))], cols: Int, rows: Int): MultibandTile = {
+    val headTile = pieces.head._1
+    val bands = Array.fill[MutableArrayTile](headTile.bandCount)(ArrayTile.empty(headTile.cellType, cols, rows))
+
+    for ((tile, (updateColMin, updateRowMin)) <- pieces) {
+      for(b <- 0 until headTile.bandCount) {
+        bands(b).update(updateColMin, updateRowMin, tile.band(b))
+      }
+    }
+
+    ArrayMultibandTile(bands)
+  }
+
+  def apply(rgbTile: MultibandTile): MultibandTile = {
+    val tileSegments = 16
```

Comment:
It turns out the weird coloring was actually caused by the relation between `uBar` and `vBar` being inverted, so the corrections were applied to the wrong channel.

I tried resampling instead of segmenting the tiles but it didn't solve the problem of inconsistent coloring, rather it applied weird corrections to the entire tile. It did, however, speed up the operation noticeably.